### PR TITLE
[WIP] add focal_loss and label_smoothing_loss

### PR DESCRIPTION
**Pull request by rohitgr7**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Added FocalLoss and LabelSmoothingLoss as discussed on slack.
TODO:
1. make changes if requested
2. update docs
3. Add tests

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
cc @justusschock

**Comment by justusschock** [comment]
@rohitgr7 can you maybe also add a functional interface?

when you look at https://github.com/justusschock/dl-utils/blob/master/dlutils/losses/focal.py you can find a more detailed interface, that also works on binary/logits. Could we maybe adopt something similar here? I'm afraid I haven't tested the implementation there. The `reduce` and `onehot` functions should be available in pytorch-lightning.metrics

**Comment by rohitgr7** [comment]
@justusschock Made the changes locally with the new interface. Any resources from where I can test my implementation. All I can find online is just the implementation of binary-version which I tested and working fine. The problem is with multi-class version one. Can't find anything for that.

**Comment by rohitgr7** [comment]
Mind the docs as of now. Tested `BinaryFocalLoss` \ `WithLogits` and `LabelSmoothingLoss`. Added `weight` and `ignore_index` parameters to match with pytorch loss api. Can't find multi-class version of focal loss to test it. Also with `weight` parameter pytorch uses weighted mean when `reduction='mean'`, but pl_metrics `reduce` function takes `elementwise_mean` which is just the basic mean. Not sure if I should keep `weight` parameter here.

**Comment by Borda** [comment]
@rohitgr7 how is it going here, is it still WIP? 🐰

**Comment by rohitgr7** [comment]
totally forgot about this one. will get this done next week.

**Comment by Borda** [comment]
@rohitgr7 any update here? 👼
